### PR TITLE
GPP-249 adjusted the chargeback notification flow

### PR DESCRIPTION
**Pull request by holkra**

### CODE REVIEW

File: src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java [hunk 1]
```diff
@@ -204,6 +203,32 @@ private void handleNotification(final NotificationRequestItem item) {
         handleNotification(null, item);
     }
 
+    private void handleChargebackNotification(final NotificationRequestItem item, boolean success) {
+        final NotificationItem notification = new NotificationItem(item);
+        final DateTime utcNow = clock.getUTCNow();
+
+        final AdyenResponsesRecord originalRecord = getResponseRecord(item.getOriginalReference());
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java]: `originalRecord` might be `null` (if the notification is for a payment not handled originally by Kill Bill).

File: src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java [hunk 2]
```diff
@@ -204,6 +203,32 @@ private void handleNotification(final NotificationRequestItem item) {
...
+        final UUID kbAccountId = UUID.fromString(originalRecord.getKbAccountId());
+        final UUID kbTenantId = UUID.fromString(originalRecord.getKbTenantId());
+        final UUID kbPaymentId = UUID.fromString(originalRecord.getKbPaymentId());
+
+        final CallContext context = new AdyenCallContext(utcNow, kbTenantId);
+        final Account account = getAccount(kbAccountId, context);
+
+        UUID kbPaymentTransactionId = null;
+        try {
+            final Payment payment = osgiKillbillAPI.getPaymentApi().createChargeback(account, kbPaymentId, notification.getAmount(), Currency.valueOf(notification.getCurrency()), null, context);
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java]: Could we populate the external key? Should it map to the `pspReference` to the `NotificationRequestItem` maybe (if it is indeed unique for that chargeback)?

File: src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java [hunk 3]
```diff
@@ -204,6 +203,32 @@ private void handleNotification(final NotificationRequestItem item) {
...
+            final PaymentTransaction lastTransaction = filterForLastTransaction(payment);
+            if (TransactionType.CHARGEBACK.equals(lastTransaction.getTransactionType())) {
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java]: Confused about that `if`. I would just filter using the `externalKey` instead.

File: src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java (hunk 1 above)
Comment (holkra) [inline src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java]: This is correct, but could that happen? And if how should we react?

File: src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java (hunk 2 above)
Comment (holkra) [inline src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java]: I populate the externalKey now. I can't use the pspReference as it is possible that we receive more than one chargeback for one payment and then it is possible that the pspReference will be used more than once.

File: src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java (hunk 3 above)
Comment (holkra) [inline src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java]: I am checking for the externalKey now, but it caused an additional class, because otherwise it wasn't possible to have the externalKey in the unit test.

File: src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java (hunk 1 above)
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java]: Yes, it can happen if two payment systems process payments for the same merchant accounts. During the migration, Adyen would be configured to send notifications to both systems.

I would just record the row and stop the flow, without throwing an exception (to prevent Adyen from retrying).

File: src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java (hunk 2 above)
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java]: >  I can't use the pspReference as it is possible that we receive more than one chargeback for one payment and then it is possible that the pspReference will be used more than once.

Adyen doesn't give us any unique reference for that particular chargeback? Populating a random UUID doesn't help here for tracking (Kill Bill will do it for you if you pass null anyways).

File: src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java (hunk 1 above)
Comment (holkra) [inline src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java]: I have implemented this and also added a unit test.

File: src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java (hunk 2 above)
Comment (holkra) [inline src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java]: No Adyen doesn't give us one. Populating the UUID as I did here, makes it possible to use the it in the test. So I would prefer to keep that here.

Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java]: > No Adyen doesn't give us one.

In that case, let's revert this change and let Kill Bill generate the UUID (in general, we prefer delegating the UUID generation to Kill Bill as much as possible, as it implements its own algorithm for performance reasons, see https://github.com/killbill/killbill/blob/master/api/src/main/java/org/killbill/billing/util/UUIDs.java).

Comment (holkra) [inline src/main/java/org/killbill/billing/plugin/adyen/core/KillbillAdyenNotificationHandler.java]: I have reverted this part.

**Comment by pierre** [comment]
Thanks!
